woocommerce/woocommerce
# Migrate E2E Test Suite from Selenium Webdriver to Puppeteer

<strike>This PR is a proof of concept to explore visual diff testing on WooCommerce core.</strike>

<hr>

**UPDATE by @juliaamosova (September 24, 2019):**

This PR is a base PR for E2E Test Suite migration from Selenium Webdriver to Puppeteer. It will serve as a switch as soon as the migration is complete. I am updating the title of this PR to reflect its purpose. 

### Changes proposed in this Pull Request:

This PR includes the following:

1) Remove Selenium Webdriver e2e test suite architecture;
2) Remove 10 Selenium Webdriver e2e tests;
3) Add Puppeteer + Jest e2e test suite architecture;
4) Add 10 Puppeteer e2e tests;
5) New `docker-compose.yaml` and `Dockerfile` in order to run e2e tests locally using Docker;
6) New `@jest/test-sequencer` package in order to run tests in a particular order;
7) New `activate-and-setup/setup-wizard.test.js` - new e2e test that will run before all other tests;
8) New scripts for running e2e tests:
- `test:e2e` - for running tests in default headless mode;
- `test:e2e-dev` - for running e2e tests in `dev` mode (headless mode off and SloMo mode on);
9) New config file `test:e2e.json` to pull variables from, needed to run e2e tests;
10) Travis implementation for running e2e tests with Docker;
11) New README file. 

More details on each point:

1) **Docker**: this PR adds Dockerfiles needed to build a test site locally using Docker. An [official WordPress Docker image](https://github.com/docker-library/docs/blob/master/wordpress/README.md#-via-docker-stack-deploy-or-docker-compose) is used to build the site.  Once the site using the WP Docker image is built, the current WooCommerce dev branch is being copied to the `plugins` folder of that newly built test site. No WooCommerce Docker image is being built or needed. 

2) **Jest test sequencer**: by default, jest runs tests ordered by the time it takes to run the test (the test that takes longer to run will be run first, the test that takes less time to run will run last). Jest sequencer introduces tools that can be used to specify the order in which the tests are being run. In our case, they are being run in alphabetical order of the directories where tests are located. This way, tests in the new directory `activate-and-setup` will run first. 

3) **Setup Wizard E2E test**: this test (located in `activate-and-setup` directory) will run before all other tests. This will allow making sure that WooCommerce is activated on the site and for the setup wizard to be completed on a brand new install of WooCommerce + it will also test Setup Wizard flow. 

4) **`test:e2e-dev` script**: this allows to run `npm run test:e2e-dev` in order to run e2e test suite in dev mode. Previously, we had to add `--dev` flag to the `test:e2e` script. 

5) **`test:e2e.json` config file**: needed so that we won't need to set test site URL, admin and customer details manually anymore. 

6) **Travis**: we can't test the Travis implementation in this PR since this branch was not made of the master. We need to merge this PR to our migration branch (https://github.com/woocommerce/woocommerce/pull/24218) which is of the master to see if the Travis implementation works and how tests run there. 

### How to test the changes in this Pull Request:

1. Install Docker Desktop if you don't have it installed:

- [Docker Desktop for Mac](https://docs.docker.com/docker-for-mac/install/)
- [Docker Desktop for Windows](https://docs.docker.com/docker-for-windows/install/)

Once installed, you should see `Docker Desktop is running` message with the green light next to it (Docker Desktop for Mac):

![docker](https://user-images.githubusercontent.com/19143190/69484051-891ac300-0e26-11ea-84ac-93a8d235173e.jpg)

**Note, that if you install docker through other methods such as homebrew, for example, your steps to set it up will be different. The commands listed in steps below may also vary.** 

2. Checkout the branch of this PR to test and stay on this branch. 

Note that by default, the local test site will be built to access on `http://localhost:8084/`. If you need to change the port (from `8084` to another one), open `docker-compose.yaml` file and change it on lines 23 and 44. 

3. Run `nmp install`

4. Run `composer install --no-dev`

5. Run `npm run build`

6. Run the following command to build the test site using Docker: `docker-compose up` and watch the site being built. Note that it may take a few minutes the first time you do that. 
7. Open new terminal window and `cd` to the current branch again.
8. Run the following command to make sure the containers were built and running: `docker ps`. You should see the 2 following containers: 

- `woocommerce_wordpress-woocomerce-dev`
- `mariadb:10.4`

9. Navigate to `http://localhost:8084/` to view the newly built site. 

10. Since we specify the site URL and create admin and customer users in docker-compose file, there is `./tests/e2e-tests/config/default.json` file that contains the variables needed to run the test:

```
{
  "url": "http://localhost:8084/",
  "users": {
    "admin": {
      "username": "admin",
      "password": "password"
    },
    "customer": {
      "username": "customer",
      "password": "password"
    }
  }
}
```

If you changed either site URL or users details in the docker-compose file, you'd need to copy the content of the `default.json`, paste it to `test:e2e.json` and edit it further there to match your own setup. 

11. We can now run the tests by running: 

- `npm run test:e2e` (headless)
- `npm run test:e2e-dev` (dev mode: non-headless with SloMo)

12. Make sure all tests are passing. 

**Note:** to skip tests, use `.only` in the relevant test entry to specify the tests that you do want to run. 

For example, in order to skip Setup Wizard tests, add `.only` to the login and activation tests as follows in the `setup-wizard.test.js`:

```
it.only( 'Can login', async () => {}
```

```
it.only( 'Can make sure WooCommerce is activated. If not, activate it', async () => {}
```

As a result, when you run `setup-wizard.test.js`, only the login and activate tests will run. The rest will be skipped. You should see the following in the terminal:

```
 PASS  tests/e2e-tests/specs/activate-and-setup/setup-wizard.test.js (11.927s)
  Store owner can login and make sure WooCommerce is activated
    ✓ Can login (7189ms)
    ✓ Can make sure WooCommerce is activated. If not, activate it (1187ms)
  Store owner can go through store Setup Wizard
    ○ skipped Can start Setup Wizard
    ○ skipped Can fill out Store setup details
    ○ skipped Can fill out Payment details
    ○ skipped Can fill out Shipping details
    ○ skipped Can fill out Recommended details
    ○ skipped Can skip Activate Jetpack section
    ○ skipped Can finish Setup Wizard - Ready! section
  Store owner can finish initial store setup
    ○ skipped Can enable tax rates and calculations
    ○ skipped Can configure permalink settings
```

<hr>

**TO STOP & RESTART DOCKER:**

- Press `Ctrl+C` in the terminal window where the containers are running 
- Stop the container(s) using the following command: `docker-compose down`
- Restart the containers using the following command: `docker-compose up`

**TO STOP DOCKER & DO A CLEAN RESTART:**

Steps below will allow building a brand new site with a clean DB and no data as it was built initially:

- Press `Ctrl+C` in the terminal window where the containers are running 
- Stop the container(s) and delete all volumes using the following command: `docker-compose down -v`
- Restart the containers using the following command: `docker-compose up --build`
